### PR TITLE
partition: Add extensive reorganize partition test

### PR DESCRIPTION
**Pull request by mjonss**
### What problem does this PR solve?

Issue Number: ref #38535

Problem Summary:

### Check List

Tests 

- [ ] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: table/tables/export_test.go [hunk 1]
```diff
@@ -0,0 +1,35 @@
+// Copyright 2023 PingCAP, Inc.
```
Comment (hawkingrei) [inline table/tables/export_test.go:1]: remember to run ```make bazel_prepare```.

File: table/tables/partition_test.go [hunk 2]
```diff
@@ -2349,3 +2356,653 @@ func TestPruneModeWarningInfo(t *testing.T) {
...
+			// update (dup in new/old , non dup as in same old/new partition -> update, different new/old -> insert + delete)
+			// delete
+			// verify with select after commit?
+
+			logutil.BgLogger().Info("State before ins/upd/del", zap.Int("transitions", transitions),
+				zap.Int("rows", len(pkMap)), zap.Stringer("SchemaState", job.SchemaState))
+			if job.SchemaState == model.StateDeleteReorganization {
+				//tk2.MustQuery(`explain format='brief' select count(distinct a) from t`).Check(testkit.Rows())
+			}
+			//tk2.MustQuery(`(select a,c from t except select a,c from t2) union (select a,0 from t except select a,0 from t2)`).Check(testkit.Rows())
```
Comment (bb7133) [inline table/tables/partition_test.go:2547]: Should we remove the commented codes?

File: table/tables/partition.go [hunk 3]
```diff
@@ -1640,9 +1643,12 @@ func partitionedTableUpdateRecord(gctx context.Context, ctx sessionctx.Context,
 			return errors.Trace(err)
 		}
 		if newTo == newFrom {
-			// Update needs to be done in StateDeleteOnly as well
 			tbl = t.GetPartition(newTo)
-			err = tbl.UpdateRecord(gctx, ctx, h, currData, newData, touched)
+			if t.Meta().Partition.DDLState == model.StateDeleteOnly {
+				err = tbl.RemoveRecord(ctx, h, currData)
```
[the remark this answers is not in the document]
Comment (tiancaiamao) [reply table/tables/partition.go:1648]: If the primary key change, `UpdateRecord` is equal to `RemoveRecord` + `AddRecord`
If the primary key is unchanged, `UpdateRecord` is `UpdateRecord`, literally @bb7133

File: table/tables/partition_test.go (hunk 2 above)
Comment (mjonss) [reply table/tables/partition_test.go:2547]: Ah, yes :) (there were left over after debugging...)

Comment (mjonss) [reply table/tables/partition_test.go]: Fixed

**Review by tiancaiamao — APPROVED** [review]

**Review by bb7133 — APPROVED** [review]
LGTM

**Review by hawkingrei — APPROVED** [review]

**Pull request merged**